### DISCUSSION TITLE
Ragged hierarchy warning sign unnatural

### DISCUSSION THREAD

**Discussion by Klimkiv** [DaxStudio/DaxStudio#672] [General]
I do test levels of a ragged hierarchy (P&L), there are not duplicates in categories and no duplicates in sorting columns. I'm choose option 'hide if blank member'
if i make hierarchy from first 2 levels where are not duplicates and blank members i've got no warning sign from dax studio
if i make hierarchy from first 3 levels where are not duplicates and are blank members on the 3rd level i've got warning sign from dax studio about unnatural hierarchy and possible performance issues
is it by design or i do something wrong?

**Comment by dgosbell** [comment]
Ragged hierarchies in tabular models will be unnatural by definition so you have not done anything wrong this is just the way things are at the moment.

> if i make hierarchy from first 3 levels where are not duplicates and are blank members on the 3rd level i've got warning sign from dax studio about unnatural hierarchy and possible performance issues

It is your blanks that will be duplicated. An unnatural hierarchy is one where any value (including blank) has more than one value in the parent column. In your case because you have a ragged hierarchy there  will be multiple different values in L2 which have blank in L3. So effectively the blank value at L3 has multiple parents at L2. 

While unnatural hierarchies can have performance issues, it mainly affects MDX clients like Excel. These issues do not affect DAX based clients like Power BI. And given that most P&L hierarchies are relatively small it's usually not a major concern.

**Reply by Klimkiv** [reply]
Darren,
Thank you very much!
One question, big bank, tabular model with balance, hierarchy with 1000 unique members, 100 mln rows fact table
If it was you what is your choice, natural or ragged? 
And one more😁
Sorting fact table before processing, should i do make it?

**Reply by dgosbell** [reply]
If you have the choice natural is better, but 1000 members is not particularly large so it does not really matter at that scale.

> Sorting fact table before processing, should i do make it?

The storage engine will test a number of different sort orders when refreshing the data anyway. Sorting really only influences the first option it tries and it's really only worth it if you think you can make a choice that will result in better compression than the optimization algorithms built in to the engine. So you'd really need to test with and without sorting to see if it has an impact. I have not worried too much about sorting fact tables so I don't have any metrics on if it's worth it or not.

**Reply by Klimkiv** [reply]
Thank you so much

**Reply by marcosqlbi** [reply]
Just adding my 2 cents.
Sorting the table really matters only if you have 100M rows or more. For smaller tables, I would not pay too much attention to that. In Power BI Desktop you might observe a difference in size because of the distribution of data between segments, but at that scale it's not something that could affect the query performance.
